### PR TITLE
spark-dataframe.md

### PR DESCRIPTION
**Pull request by Le-Zheng**

### CODE REVIEW

File: readthedocs/source/doc/UseCase/spark-dataframe.md [hunk 1]
```diff
@@ -2,4 +2,45 @@
...
+
+```python
+# update label starting from 0. That's because ratings go from 1 to 5, while the matrix column index goes from 0 to 4
+df = df.withColumn('label', df.label-1)
+
+# split to train/test dataset
+train_data, test_data = df.randomSplit([0.8, 0.2], 100)
+```
+
+### **3. Put Spark DataFrame in distributed training**
```
Comment (jason-dai) [inline readthedocs/source/doc/UseCase/spark-dataframe.md:35]: Section 3 should be define the NCF model using TF2, and Section 4 should be distributed training on Spark Dataframe

File: readthedocs/source/doc/UseCase/spark-dataframe.md [hunk 2]
```diff
@@ -2,4 +2,103 @@
...
+
+### **1. Read file into Spark DataFrame** 
+
+Spark supports to read files in CSV, JSON, TEXT, Parquet, and many more file formats into Spark DataFrame. Spark SQL provides `spark.read.csv("path")` to read a CSV file into Spark DataFrame. 
+
+```python
+from zoo.common.nncontext import *
+
+sc = init_nncontext()
+sqlcontext = SQLContext(sc)
```
Comment (jason-dai) [inline readthedocs/source/doc/UseCase/spark-dataframe.md:21]: you should use `OrcaContext.get_spark_session()` in the notebook

File: readthedocs/source/doc/UseCase/spark-dataframe.md [hunk 3]
```diff
@@ -2,4 +2,103 @@
...
+# split to train/test dataset
+train_data, test_data = df.randomSplit([0.8, 0.2], 100)
+```
+
+### **3. Define NCF model**
+
+This example defines NCF model in the _Creator Function_ using TensroFlow 2 APIs as follows.
+
+```python
+import tensorflow as tf
```
Comment (jason-dai) [inline readthedocs/source/doc/UseCase/spark-dataframe.md:42]: In the notebook, change to
```python
from tensorflow import keras
```

and then you can directly use `keras.layers` instead of `tf.keras.layers`

File: docs/readthedocs/source/doc/UseCase/spark-dataframe.md [hunk 4]
```diff
@@ -2,4 +2,100 @@
...
+
+---
+
+**In this guide we will describe how to use Spark Dataframes to scale-out data processing for distribtued deep learning.**
+
+The dataset in this guide we used is [movielens-1M](https://grouplens.org/datasets/movielens/1m/), which contains 1 million ratings of 5 levels from 6000 users on 4000 movies. We will read the data into Spark Dataframe and directly use the Spark Dataframe as the input to the distributed training.
+
+### **1. Read file into Spark DataFrame**
+
+Spark supports to read files in CSV, JSON, TEXT, Parquet, and many more file formats into Spark DataFrame. Spark SQL provides `spark.read.csv("path")` to read a CSV file into Spark DataFrame.
```
Comment (jason-dai) [inline docs/readthedocs/source/doc/UseCase/spark-dataframe.md:15]: First, read the input data into Spark Dataframe.

File: docs/readthedocs/source/doc/UseCase/spark-dataframe.md [hunk 5]
```diff
@@ -2,4 +2,100 @@
...
+from zoo.orca import OrcaContext
+
+spark = OrcaContext.get_spark_session()
+# read csv with specifying column names
+df = spark.read.csv(new_rating_files, sep=':', inferSchema=True).toDF(
+  "user", "item", "label", "timestamp")
+```
+
+### **2. Preprocess Spark DataFrame**
+
```
Comment (jason-dai) [inline docs/readthedocs/source/doc/UseCase/spark-dataframe.md:27]: Next, process the data using Spark Dataframe operations.

File: docs/readthedocs/source/doc/UseCase/spark-dataframe.md [hunk 6]
```diff
@@ -2,4 +2,100 @@
...
+    model = keras.Model(inputs=[user, item], outputs=outputs)
+    model.compile(optimizer= "adam",
+                  loss= "sparse_categorical_crossentropy",
+                  metrics=['accuracy'])
+    return model
+```
+
+### **4. Use Spark DataFrame in distributed training**
+
```
Comment (jason-dai) [inline docs/readthedocs/source/doc/UseCase/spark-dataframe.md:76]: Finally, run distributed model training/inference on the Spark Dataframes directly.

File: docs/readthedocs/source/doc/UserGuide/notebooks.md [hunk 7]
```diff
@@ -20,3 +20,8 @@
 
 ![](../../../image/colab_logo_32px.png)[Run in Google Colab](https://colab.research.google.com/github/intel-analytics/analytics-zoo/blob/master/docs/docs/colab-notebook/orca/quickstart/pytorch_lenet_mnist.ipynb) &nbsp;![](../../../image/GitHub-Mark-32px.png)[View source on GitHub](https://github.com/intel-analytics/analytics-zoo/blob/master/docs/docs/colab-notebook/orca/quickstart/pytorch_lenet_mnist.ipynb)
 
+## UseCase notebooks
+
+- **Spark Dataframes use case**
+
+![](../../../image/colab_logo_32px.png)[Run in Google Colab](https://colab.research.google.com/github/intel-analytics/analytics-zoo/blob/master/docs/docs/colab-notebook/orca/quickstart/ncf_dataframe.ipynb) &nbsp;![](../../../image/GitHub-Mark-32px.png)[View source on GitHub](https://github.com/intel-analytics/analytics-zoo/blob/master/docs/docs/colab-notebook/orca/quickstart/ncf_dataframe.ipynb)
```
Comment (jason-dai) [inline docs/readthedocs/source/doc/UserGuide/notebooks.md:27]: Change to two sections:

- Quick start
- Common use case

File: docs/docs/colab-notebook/orca/quickstart/ncf_dataframe.ipynb [hunk 8]
```diff
@@ -149,6 +149,7 @@
         "from zoo.orca import OrcaContext\r\n",
```
Comment (jason-dai) [inline docs/docs/colab-notebook/orca/quickstart/ncf_dataframe.ipynb:149]: Change the section title `NCF (data preprocessing with Spark Dataframes)` to `Using Spark Dataframes for Distribtued Deep Learning`

File: docs/readthedocs/source/doc/UseCase/spark-dataframe.md [hunk 9]
```diff
@@ -2,4 +2,106 @@
...
+```python
+from zoo.orca import OrcaContext
+
+spark = OrcaContext.get_spark_session()
+# read csv with specifying column names
+df = spark.read.csv(new_rating_files, sep=':', inferSchema=True).toDF(
+  "user", "item", "label", "timestamp")
+```
+
+### **2. Process data using Spark Dataframe operations**
```
Comment (jason-dai) [inline docs/readthedocs/source/doc/UseCase/spark-dataframe.md:25]: Process data using Spark Dataframe

File: docs/readthedocs/source/doc/UseCase/spark-dataframe.md [hunk 10]
```diff
@@ -2,4 +2,106 @@
...
+        outputs = keras.layers.Dense(units=5, activation='softmax')(concatenation)
+
+    model = keras.Model(inputs=[user, item], outputs=outputs)
+    model.compile(optimizer= "adam",
+                  loss= "sparse_categorical_crossentropy",
+                  metrics=['accuracy'])
+    return model
+```
+
+### **4. Use Spark DataFrame in distributed training**
```
Comment (jason-dai) [inline docs/readthedocs/source/doc/UseCase/spark-dataframe.md:78]: 4. Fit with Orca Estimator

**Comment by Le-Zheng** [comment]
pre-review on readthedocs website:
https://le-sparkdfguide.readthedocs.io/en/latest/doc/UseCase/spark-dataframe.html

File: docs/readthedocs/source/doc/UseCase/spark-dataframe.md [hunk 11]
```diff
@@ -2,4 +2,108 @@
...
 
-**In this guide we will describe how to use Spark Dataframes to scale-out data processing for distribtued deep learning.**
\ No newline at end of file
+![](../../../image/colab_logo_32px.png)[Run in Google Colab](https://colab.research.google.com/github/intel-analytics/analytics-zoo/blob/master/docs/docs/colab-notebook/orca/quickstart/ncf_dataframe.ipynb) &nbsp;![](../../../image/GitHub-Mark-32px.png)[View source on GitHub](https://github.com/intel-analytics/analytics-zoo/blob/master/docs/docs/colab-notebook/orca/quickstart/ncf_dataframe.ipynb)
+
+---
+
+**In this guide we will describe how to use Spark Dataframes to scale-out data processing for distribtued deep learning.**
+
+The dataset in this guide we used is [movielens-1M](https://grouplens.org/datasets/movielens/1m/), which contains 1 million ratings of 5 levels from 6000 users on 4000 movies. We will read the data into Spark Dataframe and directly use the Spark Dataframe as the input to the distributed training.
```
Comment (jason-dai) [inline docs/readthedocs/source/doc/UseCase/spark-dataframe.md:11]: The dataset used in this guide is ...

File: docs/readthedocs/source/doc/UseCase/spark-dataframe.md [hunk 12]
```diff
@@ -2,4 +2,108 @@
...
+
+### **1. Read input data into Spark DataFrame**
+
+First, read the input data into Spark Dataframes. Spark supports to read files in CSV, JSON, TEXT, Parquet, and many more file formats into Spark DataFrame. Spark SQL provides `spark.read.csv("path")` to read a CSV file into Spark DataFrame.
```
Comment (jason-dai) [inline docs/readthedocs/source/doc/UseCase/spark-dataframe.md:15]: remove `Spark supports to read files in CSV, JSON, TEXT, Parquet, and many more file formats into Spark DataFrame. Spark SQL provides `spark.read.csv("path")` to read a CSV file into Spark DataFrame.`

**Comment by jason-dai** [comment]
See comments above; others LGTM.

**Comment by Le-Zheng** [comment]
jenkins passed: http://10.239.47.210:18888/view/ZOO-PR/job/ZOO-PR-Quickstart-notebook/315
